File: webrtc/src/main/java/dev/onvoid/webrtc/media/video/desktop/DesktopFrame.java
/*
 * Copyright 2019 Alex Andres
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package dev.onvoid.webrtc.media.video.desktop;

import java.awt.Dimension;
import java.awt.Rectangle;
import java.nio.ByteBuffer;
import java.util.StringJoiner;

public class DesktopFrame {

	/**
	 * The rectangle in full desktop coordinates.
	 */
	public final Rectangle frameRect;

	/**
	 * The size of the frame in full desktop coordinate space.
	 */
	public final Dimension frameSize;

	/**
	 * The scale factor from DIPs to physical pixels of the frame.
	 */
	public final float scale;

	/**
	 * Distance in the buffer between two neighboring rows in bytes.
	 */
	public final int stride;

	/**
	 * The underlying frame buffer.
	 */
	public final ByteBuffer buffer;


	public DesktopFrame(Rectangle frameRect, Dimension frameSize, float scale, int stride, ByteBuffer buffer) {
		this.frameRect = frameRect;
		this.frameSize = frameSize;
		this.scale = scale;
		this.stride = stride;
		this.buffer = buffer;
	}

	@Override
	public String toString() {
		return new StringJoiner(", ",
				DesktopFrame.class.getSimpleName() + "[", "]")
				.add("frameRect=" + frameRect)
				.add("frameSize=" + frameSize)
				.add("scale=" + scale)
				.add("stride=" + stride)
				.add("buffer=" + buffer)
				.toString();
	}
}

File: webrtc-jni/src/main/cpp/src/media/video/desktop/DesktopFrame.cpp
/*
 * Copyright 2019 Alex Andres
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "media/video/desktop/DesktopFrame.h"
#include "JavaClasses.h"
#include "JavaDimension.h"
#include "JavaRectangle.h"
#include "JNI_WebRTC.h"

namespace jni
{
	namespace DesktopFrame
	{
		JavaLocalRef<jobject> toJava(JNIEnv * env, const webrtc::DesktopFrame * frame)
		{
			if (frame == nullptr) {
				return nullptr;
			}

			const webrtc::DesktopRect & rect = frame->rect();
			const webrtc::DesktopSize & size = frame->size();

			const auto javaClass = JavaClasses::get<JavaDesktopFrameClass>(env);

			jobject buffer = env->NewDirectByteBuffer(frame->data(), frame->stride() * frame->size().height());

			jobject object = env->NewObject(javaClass->cls, javaClass->ctor,
				JavaRectangle::toJava(env, rect.left(), rect.top(), rect.width(), rect.height()).get(),
				JavaDimension::toJava(env, size.width(), size.height()).get(),
				static_cast<jfloat>(frame->scale_factor()),
				static_cast<jint>(frame->stride()),
				buffer);

			return JavaLocalRef<jobject>(env, object);
		}

		JavaDesktopFrameClass::JavaDesktopFrameClass(JNIEnv * env)
		{
			cls = FindClass(env, PKG_DESKTOP"DesktopFrame");

			ctor = GetMethod(env, cls, "<init>", "(Ljava/awt/Rectangle;Ljava/awt/Dimension;FI" BYTE_BUFFER_SIG ")V");
		}
	}
}